R/get_pto.r

```diff
@@ -4,59 +4,118 @@
+#' @param endDate date; end date of the range for which PTO records are desired; *required*; defaults to NULL
+#' @param year int; year for which PTO records are desired; defaults to current year
+#' @param verbose logical; indicates if detailed output from httr calls should be provided; default FALSE
 #' @return tbl_df
 #'
 #' @examples
-#'
+#' \dontrun{
 #' user <- 'your_api_user'
 #' password <- 'your_password'
 #' employees <- get_pto(user=user,password=password)
+#'}
 #'
 #' @author Mark Druffel, \email{mdruffel@propellerpdx.com}
 #' @references \url{https://www.bamboohr.com/api/documentation/},  \url{https://github.com/r-lib/httr}
 #'
-#' @import httr
-#' @importFrom magrittr %>%
-#' @import purrr
-#' @import dplyr
-#' @importFrom lubridate ymd
-#' @importFrom lubridate ymd_hms
-#' @import stringr
-#' @importFrom jsonlite fromJSON
 #' @export
-get_pto <- function(user=NULL,password=NULL,verbose=FALSE){
-  time_off <- httr::GET(paste0('https://api.bamboohr.com/api/gateway.php/propellerpdx/v1/time_off/requests/?status=approved&start=',lubridate::floor_date(lubridate::today(),'year'),'&end=',lubridate::ceiling_date(lubridate::today(),'year')-1),
-                        httr::add_headers(Accept = "application/json"),
-                        httr::authenticate(user=paste0(user), password=paste0(password)),
-                        config=config(verbose=verbose)) %>%
+get_pto <- function(user=NULL,
+                    password=NULL,
+                    startDate = NULL,
+                    endDate = NULL,
+                    verbose=FALSE){
+  if(is.null(startDate) || !lubridate::is.Date(startDate)) {
+    stop("A startDate is required and must be a date!!")
+  }
+  if(is.null(endDate) || !lubridate::is.Date(endDate)) {
+    stop("A endDate is required and must be a date!!")
+  }
+
+  time_off <-
+    httr::GET(
+      paste0(
+        'https://api.bamboohr.com/api/gateway.php/propellerpdx/v1/time_off/requests/?status=approved&start=',
+        startDate,
+        '&end=',
+        endDate
+      ),
+      httr::add_headers(Accept = "application/json"),
+      httr::authenticate(user = paste0(user), password = paste0(password)),
+      config = config(verbose = verbose)
+    ) %>%
     httr::content(.,as='text',type='json',encoding='UTF-8') %>%
     jsonlite::fromJSON(.,simplifyDataFrame=F)
+
   time_off_days <- dplyr::bind_cols(
     time_off %>%
-      purrr::map(.,'dates') %>%
-      purrr::map(.,names) %>%
-      purrr::map2_df(.x=.,.y=time_off %>% purrr::map(., `[`, c('id','employeeId')), function(x=.x,y=.y) unlist(x) %>% dplyr::as_data_frame(.) %>% dplyr::mutate(id=unlist(y)[1],employeeId=unlist(y)[2])) %>% dplyr::rename('Date'='value'),
+      purrr::map(., 'dates') %>%
+      purrr::map(., names) %>%
+      purrr::map2_df(
+        .x = .,
+        .y = time_off %>% purrr::map(., `[`, c('id', 'employeeId')),
+        function(x = .x, y = .y) unlist(x) %>%
+          dplyr::as_data_frame(.) %>%
```

Comment:
```suggestion
          tibble::enframe('id') %>%
```

Reply:
Truly don't know if that is right

Reply:
I'm curious about why this is necessary if you got the same data before and after. What's motivating the change? 

I also don't know anywhere near enough about the inner workings of the tidyverse to say what's right here... @mark-druffel what's the diff between these two approaches?